### DISCUSSION TITLE
Cleanup of old notifications, revisions and activities

### DISCUSSION THREAD

**Discussion by lukas-schaetzle** [directus/directus#18297] [Feature Requests]
### Summary

Directus does currently not allow automatic or manual (via UI or webhook) deletion of old notifications, revisions (including flow logs) and activities. But I guess for most users storing all of these old items is unncessary and just clutters the database. We should provide a cleanup option to reduce the space requirements.

### Motivation

Most users proabably don't need the full history of activities, but since Directus saves every event (login, flow execution, edit, notification etc.), the space requirements for these not needed items can grow to a significant size over time. A lot of programs use log rotation to counter this problem but we don't have something like that in Directus right now.
Also, if a flow is executed very often, the log history in the sidepanel gets cluttered quite fast.

One could argue that you can already cleanup the database by executing custom SQL queries. But this feels a bit cumbersome, especially since Directus advertises with the catchword "no-code".

### Detailed Design

**Note:** Even though flow logs are also stored in the `directus_revisions` table I use the term "revisions" in this post to refer to all revision entries excluding flow execution logs.  This is because flows still have normal edit revisions and I want to distinguish between them and flow execution logs.

I suggest an (optional) automatic cleanup feature: We could add a new group in the project settings with the following fields:
1. Minimal number of flow logs to keep for every flow
2. Minimal number of revisions (excluding flow logs) to keep for every item
3. Minimal number of activities to keep
4. Minimal duration to keep items.

**For revisions, flow logs, activities:**
An item is deleted when it's older than the calculated date `$NOW - minimal duration to keep` and there are at least `minimal number of items to keep` newer items of the same type (e.g. revisions for the same item, flow logs for the same flow, ...). Activities related to saved revisions / flow logs are also saved even when they normally would have been deleted.

**For notifications:**
A notification is deleted when it's longer in the notification archive than `minimal duration to keep`. Maybe in this move it makes sense to rename "archive" to "recycle bin".

I'd imagine that the cleanup can be triggered manually via a button or webhook or an internal CRON job. The later could be configured via environment variables.

We could also think about adding a manual delete button for activities, logs, revisions etc. as this shares a lot of the logic.

### Must Have:
- Working database queries for every supported database type
- Configuration fields in the project settings (see detailed design)
- Webhook for cleanup execution

### Should Have:
- internal CRON job for cleanup execution

### Could Have:
- UI button for cleanup execution
- UI buttons for individual deletion of actitivity, flow log, revision, notification
- rename notifiaction "archive" to "recycle bin"

### Drawbacks

- It could be challenging to have working SQL queries for every supported database type (also see "unresolved questions").
- The cleanup will remove data which would probably have been important at a later time. (Therefore I suggest to configure NO automatic cleanup per default.)

### Alternatives

Impact of not doing this: People may be challenged with high storage requirements after some time or they have trouble writing SQL queries for manual cleanup.

### Adoption Strategy

Should be no breaking change. At least when we don't configure an automatic cleanup per default. However a small databse migration could be necessary to adjust a foreign key constraint (see [#13481](https://github.com/directus/directus/discussions/13481)).

### Unresolved Questions

See [#13481](https://github.com/directus/directus/discussions/13481)

**Comment by lukas-schaetzle** [comment]
For everyone using MySQL. I've just created a [gist ](https://gist.github.com/lukas-schaetzle/f93eff3d961ac595d0e1ab4be5f34536) with an SQL transaction which does the cleanup (use at your own risk!).
However, regarding the notifications it works a bit different than described here: I only check against the creation timestamp for archived notifiactions and NOT the timestamp when the notification has been archived.

**Reply by madc** [reply]
I asked some well-known AI to translate this into PostgreSQL. While it does seem to work (v10.11.2), I wouldn't use it without proper backup.
https://gist.github.com/madc/cee04370227a12127361a615d2ae8678

**Comment by rijkvanzanten** [comment]
Thoughts/feelings/concerns on my proposed first step? :) https://github.com/directus/directus/pull/18105

**Reply by rijkvanzanten** [reply]
It's effectively your 4th suggestion as a starting point: the ability to configure a max retention time for those items. The goal after that is to add a config to control max number of reversion items per parent, and max number of flow logs per flow

**Reply by lukas-schaetzle** [reply]
Oh yeah, good to know that this is already being worked on :) I'll take a look and drop my comments there when I have some remarks.

**Comment by stafyniaksacha** [comment]
Disabling them could be an option ?

**Comment by NilsBaumgartner1994** [comment]
+1 I realy appreciate this idea. I think a first step/alternative would be to have an activity auto remove schedule. Since we often dont want to have every activity logged (since GPDR).

**Reply by NilsBaumgartner1994** [reply]
As the IP Adress and User Agent are personal information (in germany and i think eu) cannot just save these information how we want. Please correct me if i am wrong.

**Reply by NilsBaumgartner1994** [reply]
Shortening the IP Adress and User Agent or removing them would be also an alternative

**Comment by ComfortablyCoding** [comment]
This is now available via https://github.com/directus/directus/pull/23410
